iotdb/src/main/java/org/apache/iotdb/db/sync/sender/FileSenderImpl.java

```diff
@@ -0,0 +1,507 @@
+  }
+
+  private String generateUUID() {
+    return Constans.SYNC_CLIENT + UUID.randomUUID().toString().replaceAll("-", "");
+  }
+
+  /**
+   * Create snapshots for valid files.
+   */
+  @Override
+  public Set<String> makeFileSnapshot(Set<String> validFiles) throws IOException {
+    Set<String> validFilesSnapshot = new HashSet<>();
+    try {
+      for (String filePath : validFiles) {
+        String snapshotFilePath = SyncUtils.getSnapshotFilePath(filePath);
+        validFilesSnapshot.add(snapshotFilePath);
+        File newFile = new File(snapshotFilePath);
+        if (!newFile.getParentFile().exists()) {
+          newFile.getParentFile().mkdirs();
+        }
+        Path link = FileSystems.getDefault().getPath(snapshotFilePath);
+        Path target = FileSystems.getDefault().getPath(filePath);
+        Files.createLink(link, target);
+      }
+    } catch (IOException e) {
+      LOGGER.error("Can not make fileSnapshot");
+      throw new IOException(e);
+    }
+    return validFilesSnapshot;
+  }
+
+  /**
+   * Transfer data of a storage group to receiver.
+   *
+   * @param fileSnapshotList list of sending snapshot files in a storage group.
+   */
+  public void syncData(Set<String> fileSnapshotList) throws SyncConnectionException {
+    try {
+      int successNum = 0;
+      for (String snapshotFilePath : fileSnapshotList) {
+        successNum++;
+        File file = new File(snapshotFilePath);
+        List<String> filePathSplit = new ArrayList<>();
+        String os = System.getProperty("os.name");
+        if (os.toLowerCase().startsWith("windows")) {
+          String[] name = snapshotFilePath.split(File.separator + File.separator);
+          filePathSplit.add(name[name.length - 2]);
+          filePathSplit.add(name[name.length - 1]);
+        } else {
+          String[] name = snapshotFilePath.split(File.separator);
+          filePathSplit.add(name[name.length - 2]);
+          filePathSplit.add(name[name.length - 1]);
+        }
+        while (true) {
+          // Sync all data to receiver
+          byte[] buffer = new byte[Constans.DATA_CHUNK_SIZE];
+          int data;
+          try (FileInputStream fis = new FileInputStream(file)) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(Constans.DATA_CHUNK_SIZE);
+            while ((data = fis.read(buffer)) != -1) { // cut the file into pieces to send
+              bos.write(buffer, 0, data);
+              ByteBuffer buffToSend = ByteBuffer.wrap(bos.toByteArray());
+              bos.reset();
+              serviceClient
+                  .syncData(null, filePathSplit, buffToSend, SyncDataStatus.PROCESSING_STATUS);
+            }
+            bos.close();
+          }
+
+          // Get md5 of the file.
+          MessageDigest md = MessageDigest.getInstance("MD5");
+          try (FileInputStream fis = new FileInputStream(file)) {
+            while ((data = fis.read(buffer)) != -1) {
+              md.update(buffer, 0, data);
```

Comment:
Can not the MD5 calculation be placed in the previous loop?

Reply:
Except for the last loop, the receiver doesn't calculate MD5 of the file because it's meaningless. So  it needs the last loop to notify receiver  to generate MD5

Reply:
Yeah, but in most cases (supposing the fault rate is low) the calculation of MD5 is unavoidable. If you calculate it while you are sending the file, you can avoid reading the whole file again, which can be costly if the file is large.

Reply:
I misunderstand your point before, that's a good improvement.

Reply:
fixed